shell/platform/linux/fl_application.cc

```diff
@@ -0,0 +1,183 @@
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_view.h"
+
+struct FlApplicationPrivate {
+  // Default window title to use.
+  gchar* window_title;
+
+  // Arguments to pass to Dart.
+  gchar** dart_entrypoint_arguments;
+
+  // Default width of a Flutter window in pixels.
+  int window_width;
+
+  // Default height of a Flutter window in pixels.
+  int window_height;
+};
+
+#define FL_APPLICATION_GET_PRIVATE(app)                        \
+  ((FlApplicationPrivate*)fl_application_get_instance_private( \
+      FL_APPLICATION(app)))
+
+enum { kSignalRegisterPlugins, kSignalLastSignal };
+
+static guint fl_application_signals[kSignalLastSignal];
+
+G_DEFINE_TYPE_WITH_CODE(FlApplication,
+                        fl_application,
+                        GTK_TYPE_APPLICATION,
+                        G_ADD_PRIVATE(FlApplication))
+
+// Default implementation of FlApplication::register_plugins
+static void fl_application_register_plugins(FlApplication* self,
+                                            FlPluginRegistry* registry) {}
+
+// Implements GApplication::activate.
+static void fl_application_activate(GApplication* application) {
+  FlApplication* self = FL_APPLICATION(application);
+  FlApplicationPrivate* priv = FL_APPLICATION_GET_PRIVATE(self);
+
+  GtkApplicationWindow* window =
+      GTK_APPLICATION_WINDOW(gtk_application_window_new(GTK_APPLICATION(self)));
+
+  // Use a header bar when running in GNOME as this is the common style used
+  // by applications and is the setup most users will be using (e.g. Ubuntu
+  // desktop).
+  // If running on X and not using GNOME then just use a traditional title bar
+  // in case the window manager does more exotic layout, e.g. tiling.
+  // If running on Wayland assume the header bar will work (may need changing
+  // if future cases occur).
+  gboolean use_header_bar = TRUE;
+#ifdef GDK_WINDOWING_X11
+  GdkScreen* screen = gtk_window_get_screen(GTK_WINDOW(window));
+  if (GDK_IS_X11_SCREEN(screen)) {
+    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
+    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
+      use_header_bar = FALSE;
+    }
+  }
+#endif
+  if (use_header_bar) {
+    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
+    gtk_widget_show(GTK_WIDGET(header_bar));
+    gtk_header_bar_set_title(header_bar, priv->window_title);
+    gtk_header_bar_set_show_close_button(header_bar, TRUE);
+    gtk_window_set_titlebar(GTK_WINDOW(window), GTK_WIDGET(header_bar));
+  } else {
+    gtk_window_set_title(GTK_WINDOW(window), priv->window_title);
+  }
+
+  gtk_window_set_default_size(GTK_WINDOW(window), priv->window_width,
+                              priv->window_height);
+  gtk_widget_show(GTK_WIDGET(window));
+
+  g_autoptr(FlDartProject) project = fl_dart_project_new();
+  fl_dart_project_set_dart_entrypoint_arguments(
+      project, priv->dart_entrypoint_arguments);
+
+  FlView* view = fl_view_new(project);
```

Comment:
In a multi-window world, we'll need to share an `FlEngine` across multiple `FlView`s. How will that be done? Will we introduce a `fl_application_set_engine` API?

Reply:
I think the FlApplication will create and own the FlEngine in this case, so it should be transparent to the app.

Reply:
Ah. If you have multiple windows, will you have only a single `FlApplication`?

If yes, would you mind sharing what the code for app with two windows would look like, similar to the code in the PR description above?

Reply:
Thanks! One thing that's not clear to me: how does the app dictate that it has two windows?

Reply:
There's currently no mechanism to request new windows. So the standard behaviour is you make an FlApplication, and that notifies you when new windows are created via the create-window signal. I'd expect the windows to be initiated in the Dart code, and FlEngine/FlApplication to handle this (probably via a channel) and notify the native code via the signal.

If the native side wanted to create a new window I expect this will be done by sending a request to the Dart code to do so (via a channel). The window will then be created by the same mechanism described above.

In the case of native (i.e. GTK) windows you can just add these yourself with `gtk_application_window_new(GTK_APPLICATION(app))`. Dart will be unaware of these.